src/main/java/com/netflix/imflibrary/st2067_2/Application2E2021.java

```diff
@@ -269,13 +271,266 @@ public Application2E2021(@Nonnull IMFCompositionPlaylistType imfCompositionPlayl
+        if ((maxSz <= 2048 && p.cod.numDecompLevels > 5) ||
+            (maxSz <= 4096 && p.cod.numDecompLevels > 6) ||
+            (maxSz <= 8192 && p.cod.numDecompLevels > 7)) {
+            logger.addError(
+                IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                "APP2.HT: Invalid number of decomposition levels");
+            isValid = false;
+        }
+
+        /* number of layers */
+
+        if (p.cod.numLayers != 1) {
+            logger.addError(
+                IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                String.format("APP2.HT: Number of layers (%d) is not 1", p.cod.numLayers));
+            isValid = false;
+        }
+
+        /* code-block sizes */
+
+        if (p.cod.ycb < 5 || p.cod.ycb > 6) {
+            logger.addError(
+                IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                String.format("APP2.HT: Invalid vertical code-block size (ycb = %d)", p.cod.ycb));
+            isValid = false;
+        }
+
+        if (p.cod.xcb < 5 || p.cod.xcb > 7) {
+            logger.addError(
+                IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                String.format("APP2.HT: Invalid horizontal code-block size (xcb = %d)", p.cod.xcb));
+            isValid = false;
+        }
+
+
+        /* transformation */
+
+        boolean isReversibleFilter = (p.cod.transformation == 1);
+
+        if (isHTREV && !isReversibleFilter) {
+            logger.addError(
+                IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                "APP2.HT: 9-7 irreversible filter is used but HTREV is signaled in CAP");
+            isValid = false;
+        }
+
+        /* precinct size */
+
+        if (p.cod.precinctSizes.length == 0 || p.cod.precinctSizes[0] != 0x77) {
+            logger.addError(
+                IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                "APP2.HT: Invalid precinct sizes");
+            isValid = false;
+        }
+
+        for (int i = 1; i < p.cod.precinctSizes.length; i++)
+            if (p.cod.precinctSizes[i] != 0x88) {
+                logger.addError(
+                    IMFErrorLogger.IMFErrors.ErrorCodes.APPLICATION_COMPOSITION_ERROR,
+                    IMFErrorLogger.IMFErrors.ErrorLevels.FATAL,
+                    "APP2.HT: Invalid precinct sizes");
+                isValid = false;
+                break;
+            }
+
+        /* magbp */
+
+
+        int maxB = p.csiz[0].ssiz + 2;
```

Comment:
I'd suggest adding a comment, explaining that the maxB value, is based on values in `Table I.2` and `Table I.3`.

The maxB calculations are a little hard to follow, since ST 2067-21 doesn't present these as a "formula", but rather just a table of hardcoded values.
What you are doing makes sense after reading through the code several times. It looks like you've got all the values correct. But adding a comment explaining the calculations might make this code easier for others to understand.

Reply:
Will do.

The code needs additional comments, which I plan to add once we are somewhat happy with the results and the overall flow/structure.